### PR TITLE
APP-37126 introduction of error reporting capability

### PR DESCRIPTION
**Pull request by jbornemann**
This change is [Reviewable](https://reviewable.io/reviews/pendo-io/appwrap/117)

### CODE REVIEW

**Comment by jbornemann** [comment]
Going to be testing this out on a project, but wanted to open for vis.

File: logging_mock.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package appwrap
+
+import (
+	"github.com/stretchr/testify/mock"
+	"net/http"
+)
+
+type LoggingMock struct {
+	mock.Mock
+	Log Logging // all log calls are passed through to this logger (NullLogger may be used)
```
Comment (jbornemann) [inline logging_mock.go:10]: Happy to move appengine usage to this, since it is _mostly_ shamelessly copied from appengine.

File: go.mod [hunk 2]
```diff
@@ -6,6 +6,7 @@ require (
 	cloud.google.com/go v0.57.0
 	cloud.google.com/go/datastore v1.2.0
 	cloud.google.com/go/logging v1.0.0
+	cloud.google.com/go/pubsub v1.3.1 // indirect
```
Comment (jbornemann) [inline go.mod:9]: GoLand, **really** wanted to add this. I think it will be fine?

**Review by jshore1296 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pendo-io/appwrap/117)* status: 0 of 16 files reviewed, 5 unresolved discussions (waiting on @jbornemann, @jshore1296, and @sgp)

___
*[error_reporting.go, line 6 at r1](https://reviewable.io/reviews/pendo-io/appwrap/117#-MI4Qy9oBPupTiga_XTY:-MI4Qy9oBPupTiga_XTZ:babeio) ([raw file](https://github.com/pendo-io/appwrap/blob/397007670fecc899aa4072062ecbcb37ff6234ff/error_reporting.go#L6)):*
> ```Go
> 	"cloud.google.com/go/errorreporting"
> 	"context"
> 	"fmt"
> ```

sort imports (everywhere, but I'll just make the comment once :grin: )
___
*[error_reporting.go, line 32 at r1](https://reviewable.io/reviews/pendo-io/appwrap/117#-MI4R5qT0L6CaY7SENPz:-MI4R5qT0L6CaY7SENQ-:b-9t2rzw) ([raw file](https://github.com/pendo-io/appwrap/blob/397007670fecc899aa4072062ecbcb37ff6234ff/error_reporting.go#L32)):*
> ```Go
> 
> func NewGoogleErrorReporter(info AppengineInfo, ctx context.Context, log Logging) (*googleErrorReporter, error) {
> 	client, err := errorreporting.NewClient(ctx, info.DataProjectID(), errorreporting.Config{
> ```

Are these clients meant to be long-lived (or do they need to be closed when they're done)?  If so, I'd do something similar to what we do for datastore where we have a single client that's shared among all instances of this interface.  Otherwise we'll end up with resource leaks if we don't close the gazillions of copies we make.
___
*[error_reporting.go, line 133 at r1](https://reviewable.io/reviews/pendo-io/appwrap/117#-MI4RdYI0ng5lb2Dteff:-MI4RdYI0ng5lb2Dtefg:b-5h9ztu) ([raw file](https://github.com/pendo-io/appwrap/blob/397007670fecc899aa4072062ecbcb37ff6234ff/error_reporting.go#L133)):*
> ```Go
> }
> 
> func (g *googleErrorReporter) WrapLogger(logging Logging, errorAffectsLabel string) errorForwardingLogger {
> ```

nit: we might want to check that we're not wrapping another errorForwardingLogger here.

**Comment by jbornemann** [comment]
> > _[Reviewable](https://reviewable.io/reviews/pendo-io/appwrap/117)_ status: 0 of 16 files reviewed, 5 unresolved discussions (waiting on @jbornemann, @jshore1296, and @sgp)
> 
> _[error_reporting.go, line 6 at r1](https://reviewable.io/reviews/pendo-io/appwrap/117#-MI4Qy9oBPupTiga_XTY:-MI4Qy9oBPupTiga_XTZ:babeio) ([raw file](https://github.com/pendo-io/appwrap/blob/397007670fecc899aa4072062ecbcb37ff6234ff/error_reporting.go#L6)):_
> 
> > ```go
> > 	"cloud.google.com/go/errorreporting"
> > 	"context"
> > 	"fmt"
> > ```
> 
> sort imports (everywhere, but I'll just make the comment once 😁 )
> 

thanks 😄 will do
> _[error_reporting.go, line 32 at r1](https://reviewable.io/reviews/pendo-io/appwrap/117#-MI4R5qT0L6CaY7SENPz:-MI4R5qT0L6CaY7SENQ-:b-9t2rzw) ([raw file](https://github.com/pendo-io/appwrap/blob/397007670fecc899aa4072062ecbcb37ff6234ff/error_reporting.go#L32)):_
> 
> > ```go
> > func NewGoogleErrorReporter(info AppengineInfo, ctx context.Context, log Logging) (*googleErrorReporter, error) {
> > 	client, err := errorreporting.NewClient(ctx, info.DataProjectID(), errorreporting.Config{
> > ```
> 
> Are these clients meant to be long-lived (or do they need to be closed when they're done)? If so, I'd do something similar to what we do for datastore where we have a single client that's shared among all instances of this interface. Otherwise we'll end up with resource leaks if we don't close the gazillions of copies we make.

The reporter + client can be long-lived. The way I have it now is that the entries will be sent async, and automatically flushed by the bundler after either a period of time has elapsed, or a number of entries is exceed. A final flush would probably be a nice cleanup task on server shutdown to make sure we don't loose any events. 

The loggers created by the reporter in our usage are likely short-lived - spanning the life of the request; but they can be created and utilize the same underlying reporter. 
> 
> _[error_reporting.go, line 133 at r1](https://reviewable.io/reviews/pendo-io/appwrap/117#-MI4RdYI0ng5lb2Dteff:-MI4RdYI0ng5lb2Dtefg:b-5h9ztu) ([raw file](https://github.com/pendo-io/appwrap/blob/397007670fecc899aa4072062ecbcb37ff6234ff/error_reporting.go#L133)):_
> 
> > ```go
> > }
> > 
> > func (g *googleErrorReporter) WrapLogger(logging Logging, errorAffectsLabel string) errorForwardingLogger {
> > ```
> 
> nit: we might want to check that we're not wrapping another errorForwardingLogger here.

Good call

**Review by sgp — APPROVED** [review]
> Reviewed 13 of 16 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/pendo-io/appwrap/117)* status: all files reviewed, 5 unresolved discussions (waiting on @jshore1296)

**Comment by jbornemann** [comment]
I got this hooked up in ops-webhooks, and things are looking good. So taking off the wip label

**Review by jshore1296 — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/pendo-io/appwrap/117)* status: all files reviewed, 4 unresolved discussions (waiting on @jbornemann and @jshore1296)

___
*[error_reporting.go, line 32 at r1](https://reviewable.io/reviews/pendo-io/appwrap/117#-MI4R5qT0L6CaY7SENPz:-MIz5dVb5SfBMP89vG9T:b2k2v2a) ([raw file](https://github.com/pendo-io/appwrap/blob/397007670fecc899aa4072062ecbcb37ff6234ff/error_reporting.go#L32)):*
<details><summary><i>Previously, jbornemann (Jeff Bornemann) wrote…</i></summary><blockquote>

The reporter + client can be long-lived. The way I have it now is that the entries will be sent async, and automatically flushed by the bundler after either a period of time has elapsed, or a number of entries is exceed. A final flush would probably be a nice cleanup task on server shutdown to make sure we don't loose any events. 

The loggers created by the reporter in our usage are likely short-lived - spanning the life of the request; but they can be created and utilize the same underlying reporter.  
</blockquote></details>

I mean the error reporting client itself.  If we create one error reporter per request, we're going to run out of resources very quickly. 
___
*[logging_mock.go, line 10 at r1](https://reviewable.io/reviews/pendo-io/appwrap/117#-MI0xFfPynR2eIplSCLd-r1-10:-MIz5p4B3yIjhbp3jfI5:b-8fx2u7) ([raw file](https://github.com/pendo-io/appwrap/blob/397007670fecc899aa4072062ecbcb37ff6234ff/logging_mock.go#L10)):*
<details><summary><i>Previously, jbornemann (Jeff Bornemann) wrote…</i></summary><blockquote>

Happy to move appengine usage to this, since it is _mostly_ shamelessly copied from appengine. 
</blockquote></details>

:+1:

**Review by jshore1296 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pendo-io/appwrap/117)* status: all files reviewed, 4 unresolved discussions (waiting on @jbornemann and @jshore1296)

___
*[error_reporting.go, line 32 at r1](https://reviewable.io/reviews/pendo-io/appwrap/117#-MI4R5qT0L6CaY7SENPz:-MIz6WDw3Py_XM4Nw82b:b5c1lw6) ([raw file](https://github.com/pendo-io/appwrap/blob/397007670fecc899aa4072062ecbcb37ff6234ff/error_reporting.go#L32)):*
<details><summary><i>Previously, jshore1296 wrote…</i></summary><blockquote>

I mean the error reporting client itself.  If we create one error reporter per request, we're going to run out of resources very quickly. 
</blockquote></details>

Following up from slack - can you expose the Close() method on this client, assuming there is one, and then put a comment on the constructor that these clients are meant to be long-lived and should not be created per-request?

**Review by jshore1296 — APPROVED** [review]
> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/pendo-io/appwrap/117)* status: all files reviewed, 2 unresolved discussions (waiting on @jbornemann)

**Review by jshore1296 — APPROVED** [review]
[merge]

> *[Reviewable](https://reviewable.io/reviews/pendo-io/appwrap/117#-:-MIzSyi-2Jlmfkvwb68V:b3sq3yz)* status: all files reviewed, 2 unresolved discussions (waiting on @jbornemann)

**Comment by pendo-ops** [comment]
All JIRA issue transitioned/labeled.
* [APP-37126](https://pendo-io.atlassian.net/browse/APP-37126): Jira Issue APP-37126: Changed to Resolved


Branch jb-app-37216-error-reporting deleted

**Pull request merged**